contrib/ssh-copy-id

```diff
@@ -295,9 +299,22 @@ case "$REMOTE_VERSION" in
     populate_new_ids 0
     # in ssh below - to defend against quirky remote shells: use 'exec sh -c' to get POSIX;
     #     'cd' to be at $HOME; add a newline if it's missing; and all on one line, because tcsh.
-    [ "$DRY_RUN" ] || printf '%s\n' "$NEW_IDS" | \
+    if ! [ "$DRY_RUN" ] && ! [ "$USE_SFTP" ]; then
+      printf '%s\n' "$NEW_IDS" | \
       ssh "$@" "exec sh -c 'cd ; umask 077 ; mkdir -p .ssh && { [ -z "'`tail -1c .ssh/authorized_keys 2>/dev/null`'" ] || echo >> .ssh/authorized_keys ; } && cat >> .ssh/authorized_keys || exit 1 ; if type restorecon >/dev/null 2>&1 ; then restorecon -F .ssh .ssh/authorized_keys ; fi'" \
       || exit 1
+    fi
+    if ! [ "$DRY_RUN" ] && [ "$USE_SFTP" ] ; then
+      ssh -f -N -M -S /tmp/ssh-copy-id-shared-connection "$@"
+      # todo: create .ssh/authorized_keys if not exist
+      # todo: grep the output of the sftp-commands and exit 1 on failures
+      echo "get .ssh/authorized_keys /tmp/ssh-copy-id_authorized_keys" | sftp -o "ControlPath=/tmp/ssh-copy-id-shared-connection" "notnecessary"
```

Comment:
The control path can not be this specific, since it would be very racy especially in multiuser systems or when copying keys to multiple systems "at once". It should be placed in some "random" filename/directory to avoid making it accessible by others and this race conditions.

Additionally, you are overriding this option blindly, if it was previously configured, so I would probably add some check (for example using `ssh -G`) to verify that this is not already set and use this one-shot value only if not.

The sftp line can be also written as `sftp "$@:.ssh/authorized_keys" /tmp/ssh-copy-id_authorized_keys `

And again, using the too-specific path `/tmp/ssh-copy-id_authorized_keys` for this critical file is also very racy -- it should be placed again in some temporary directory with random name.

Reply:
both should be done with mktemp

Reply:
> Additionally, you are overriding this option blindly, if it was previously configured, so I would probably add some check (for example using `ssh -G`) to verify that this is not already set and use this one-shot value only if not.

I'm not sure if we should go for every eventuality. This works for the default-case and does not conflict with this setting otherwise. The only this is that a user opens a connection with another key and wants to reuse it here. Also: some configurations are also overridden [lines above](https://github.com/openssh/openssh-portable/blob/05c215de8d224e094a872d97d45f37f60c06206b/contrib/ssh-copy-id#L239)
 
> The sftp line can be also written as `sftp "$@:.ssh/authorized_keys" /tmp/ssh-copy-id_authorized_keys `

this could result in `sftp -p 23 example.com:.ssh/authorized_keys /tmp/ssh-copy-id_authorized_keys` which will fail because sftp doesn't understand `-p`. To change this we would have to filter out the `-p` option and handle it separately for the ssh and sftp-commands.